src/Action/Server.hs

```diff
@@ -161,72 +178,86 @@ replyServer log local links haddock store cdn home htmlDir scope Input{..} = cas
         pure $ OutputFile $ joinPath $ htmlDir : xs
     where
         html = templateMarkup
-        text = templateMarkup . H.string
+        txt = templateMarkup . H.string
 
         tagOptions sel = mconcat [H.option Text.Blaze.!? (x `elem` sel, H.selected "selected") $ H.string x | x <- completionTags store]
         params =
-            [("cdn", text cdn)
-            ,("home", text home)
-            ,("jquery", text $ if null cdn then "plugin/jquery.js" else "https:" ++ JQuery.url)
-            ,("version", text $ showVersion version ++ " " ++ showUTCTime "%Y-%m-%d %H:%M" spawned)]
+            [("cdn", txt cdn)
+            ,("home", txt home)
+            ,("jquery", txt $ if null cdn then "plugin/jquery.js" else "https:" ++ JQuery.url)
+            ,("version", txt $ showVersion version ++ " " ++ showUTCTime "%Y-%m-%d %H:%M" spawned)]
         templateIndex = templateFile (htmlDir </> "index.html") `templateApply` params
         templateEmpty = templateFile (htmlDir </>  "welcome.html")
-        templateHome = templateIndex `templateApply` [("tags",html $ tagOptions []),("body",templateEmpty),("title",text "Hoogle"),("search",text ""),("robots",text "index")]
+        templateHome = templateIndex `templateApply` [("tags",html $ tagOptions []),("body",templateEmpty),("title",txt "Hoogle"),("search",txt ""),("robots",txt "index")]
         templateLog = templateFile (htmlDir </> "log.html") `templateApply` params
         templateLogJs = templateFile (htmlDir </> "log.js") `templateApply` params
 
 
-dedupeTake :: Ord k => Int -> (v -> k) -> [v] -> [[v]]
-dedupeTake n key = f [] Map.empty
+-- | Take from the list until we’ve seen `n` different keys,
+-- and group all values by their respective key.
+--
+-- Will keep the order of elements for each key the same.
+takeAndGroup :: Ord k => Int -> (v -> k) -> [v] -> [[v]]
+takeAndGroup n key = f [] Map.empty
     where
-        -- map is Map k [v]
-        f res mp xs | Map.size mp >= n || null xs = map (reverse . (Map.!) mp) $ reverse res
-        f res mp (x:xs) | Just vs <- Map.lookup k mp = f res (Map.insert k (x:vs) mp) xs
-                        | otherwise = f (k:res) (Map.insert k [x] mp) xs
+        -- mp is Map k [v]
+        f keys mp []
+            = map (\k -> reverse $ mp Map.! k) $ reverse keys
+        f keys mp _ | Map.size mp >= n
+            = map (\k -> reverse $ mp Map.! k) $ reverse keys
+        f keys mp (x:xs)
+            | Just vs <- Map.lookup k mp = f keys (Map.insert k (x:vs) mp) xs
+            | otherwise = f (k:keys) (Map.insert k [x] mp) xs
             where k = key x
 
+data UrlOpts = IsHaddockUrl | IsLocalUrl | IsOtherUrl
```

Comment:
Would be great if this data type had some comments as to what each of the types of url looked like or represented (I appreciate this was missing before, but now we have the type, it should be easy to add)

Reply:
I don’t actually know what these mean semantically, I just refactored. I’ll leave it up to you to document the cases :)